File: CH1/ch1-7.cpp
// Problem 1.7: If an entry in a matrix is 0, set the row and col to 0s

//
//  main.cpp
//  Coding practice
//
//  Created by Tiantian Xu on 6/11/16.
//  Copyright © 2016 Tiantian Xu. All rights reserved.
//

#include <iostream>
using namespace std;

void makezeros(int arr[][5],int M, int N) {
    
    bool *rows = new bool(M);
    for (int i=0;i<M;i++) rows[i]=false;
    
    bool *cols = new bool(N);
    for (int i=0;i<N;i++) cols[i]=false;
    
    for (int i=0; i<M; i++)
        for(int j=0;j<N; j++){
            if (arr[i][j]==0) {
                rows[i]=true;
                cols[j]=true;
            }
        }
    
    for (int i=0;i<M;i++)
        for (int j=0; j<N; j++) {
            if (rows[i] || cols[j])
                arr[i][j]=0;
        }
}

int main(int argc, const char * argv[]) {
    int arr[][5] = {{1,2,3,0,5},{6,7,8,9,10},{11,12,13,14,15},{0,17,18,19,20},{21,22,23,24,25}};
    int N = 5;
    
    cout<<"Originally the array is\n";
    
    for (int i=0;i<N;i++) {
        
        for(int j=0;j<N;j++)
            cout<<arr[i][j]<< " ";
        cout<<endl;
    }
    
    makezeros(arr,N,N); // arr is not necessarily square
    
    cout<<"After rotation the array is\n";
    
    for (int i=0;i<N;i++) {
        
        for(int j=0;j<N;j++)
            cout<<arr[i][j]<< " ";
        cout<<endl;
    }
    return 0;
}

File: CH1/ch1-7.py
# programmed by Hossein Siadati
def rowcolumn_zero(matrix):

	rows = [False] * len(matrix)
	cols = [False] * len(matrix[0])

	for i in range(len(rows)):
		for j in range(len(cols)):
			if matrix[i][j] == 0:
				rows[i] = True
				cols[j] = True
	
	for i in range(len(rows)):
                for j in range(len(cols)):
			if rows[i]==True or cols[j]==True:
				matrix[i][j] = 0
	return matrix

def test1():
	matrix = [[0,2,3,4],
		[6,5,7,8],
		[9,10,0,12]]
	print rowcolumn_zero(matrix)

if __name__ == "__main__":
	test1()
